### PR TITLE
xtest: add --install-tas argument

### PR DESCRIPTION
**Pull request by jenswi-linaro**
Adds --install-tas argument which will install all bootstrap TAs
(/lib/optee_armtz/*.bsta) in the OP-TEE TA database.

Signed-off-by: Jens Wiklander <jens.wiklander@linaro.org>

Needed by https://github.com/OP-TEE/optee_os/pull/1928 to install bootstrap TAs.

### CODE REVIEW

**Comment by jenswi-linaro** [comment]
Thanks @jbech-linaro, so there's a mutual dependency between the two PRs.

File: host/xtest/xtest_main.c [hunk 1]
```diff
@@ -65,6 +66,7 @@ void usage(char *program)
 	printf("applets:\n");
 	printf("\t--sha-perf [opts]  SHA performance testing tool (-h for usage)\n");
 	printf("\t--aes-perf [opts]  AES performance testing tool (-h for usage)\n");
+	printf("\t--install-tas [opts] Install TAs (-h for usage)\n");
```
Comment (etienne-lms) [inline host/xtest/xtest_main.c]: minor: remove `[opts]` and `(-h for usage)`

File: host/xtest/install_tas.c [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+}
+
+int install_tas_runner_cmd_parser(int argc __unused, char *argv[] __unused)
+{
+	TEEC_Result res;
+	uint32_t err_origin;
+	TEEC_UUID uuid = PTA_MANAGEMENT_UUID;
+	TEEC_Context ctx;
+	TEEC_Session sess;
+	const char *ta_dir = "/lib/optee_armtz";
```
Comment (etienne-lms) [inline host/xtest/install_tas.c]:
```
#ifndef TA_DIR
# ifdef __ANDROID__
#define TA_DIR "/system/lib/optee_armtz"
# else
#define TA_DIR "/lib/optee_armtz"
# endif
#endif
```
-> see **regression_1000.c**
`const char *ta_dir = TA_DIR;`

**Comment by jenswi-linaro** [comment]
Update

**Comment by jenswi-linaro** [comment]
Update to pass all test together with OP-TEE/optee_os#1928

**Comment by jforissier** [comment]
It would be more flexible as `xtest --install-ta <ta-path>`, because `TA_DIR` is not relevant for secstor TAs, and you may not want to install all TAs in a given directory.

**Comment by jforissier** [comment]
In noticed that if I install the TAs several times, more objects keep being added into the secure storage (`/data/tee/<some_number>.ta`). Is that expected?

**Comment by jenswi-linaro** [comment]
Thanks
I'll update the `--install-ta` command
It's not expected that objects should be added all the time, I'll fix.

**Comment by jenswi-linaro** [comment]
Update

File: host/xtest/install_ta.c [hunk 3]
```diff
@@ -91,24 +132,21 @@ int install_tas_runner_cmd_parser(int argc __unused, char *argv[] __unused)
...
 			continue;
-
-		printf("Installing \"%s\"\n", dent->d_name);
-		ta = read_ta(ta_dir, dent->d_name, &ta_size);
-		install_ta(&sess, ta, ta_size);
+		}
+
+		if (S_ISDIR(sb.st_mode))
+			install_dir(&sess, argv[i]);
+		else if (S_ISREG(sb.st_mode))
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply host/xtest/install_ta.c]: That's why I'm using `stat()` instead of `lstat()`

**Comment by jenswi-linaro** [comment]
If "legacy" testing it be kept in 1008, what are supposed to do when the uuid of the TA we're testing with is installed in the tadb and the TA isn't fetch from REE FS TA storage?

File: host/xtest/install_ta.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (jforissier) [reply host/xtest/install_ta.c]: Ah, right, I missed that. It's fine then.

**Comment by jenswi-linaro** [comment]
Update

**Comment by jforissier** [comment]
We also need `--uninstall-ta`. I can propose some code if you want. `--list-ta` would be good also but maybe not as important.

**Comment by jenswi-linaro** [comment]
Why is the legacy testing so important, in the end the same code is tested (core/crypto/signed_hdr.c)?

PTA_SECSTOR_TA_MGMT currently only supports installing TAs, not listing or uninstalling. I'm not sure that will ever be exposed directly to normal world either.

**Comment by jforissier** [comment]
> Why is the legacy testing so important, in the end the same code is tested (core/crypto/signed_hdr.c)?

Maybe it's not.

> PTA_SECSTOR_TA_MGMT currently only supports installing TAs, not listing or uninstalling. I'm not sure that will ever be exposed directly to normal world either.

OK, but an uninstall option is still useful for testing. Case in point:
```
root@HiKey960:/ U=cb3e5ba0-adf1-11e0-998b-0002a5d5c51b; \
  for i in `seq 1 1500`; do \
    echo >&2 $i && \
    xtest --install-ta /lib/optee_armtz/$U.ta && \
    xtest --uninstall-ta $U; \
    if [ $? -ne 0 ]; then \
      break; \
    fi; \
  done >/dev/null
[...]
1006
1007
1008
1009
xtest: install_ta: TEEC_InvokeCommand: 0xffff000c err_origin 0x4
```

This tells me we have a memory leak somewhere :(

**Comment by jbech-linaro** [comment]
> This tells me we have a memory leak somewhere :(

Try it using Valgrind, it has helped me look at various OP-TEE components in the past, including xtest (If you have no rootfs including it, let me know and I can share something with you).

**Comment by jforissier** [comment]
Valgrind on the TEE core? ;-)

**Comment by jbech-linaro** [comment]
> Valgrind on the TEE core? ;-)

Ah, sorry, that could be harder :sweat_smile: ... I was reading it as you suspected a memory leak from xtest, but that is of course not the case here.

**Comment by jenswi-linaro** [comment]
Memory leak plugged in #1928

**Comment by jenswi-linaro** [comment]
Next step is to add a test TA handles a security domain. That TA will be able to uninstall TAs etc. The reason I don't want it directly available to normal world is that it would be a nice security hole that would be enabled with testing enabled or something.

**Comment by jforissier** [comment]
> Memory leak plugged in #1928

Confirmed.

> [uninstall] The reason I don't want it directly available to normal world is that it would be a nice security hole that would be enabled with testing enabled or something.

As long as we have a simple way to exercise the install/uninstall code, it's fine.

**Comment by jenswi-linaro** [comment]
> As long as we have a simple way to exercise the install/uninstall code, it's fine.

We will not have it in this PR, but hopefully the next.

**Comment by jforissier** [comment]
@jenswi-linaro sure, that was my understanding. Thanks.

File: host/xtest/xtest_test.h [hunk 4]
```diff
@@ -122,4 +122,20 @@ extern const TEEC_UUID socket_ta_uuid;
...
 
+#ifndef TA_DIR
+# ifdef __ANDROID__
+#define TA_DIR "/system/lib/optee_armtz"
+# else
+#define TA_DIR "/lib/optee_armtz"
+# endif
+#endif
+
+#ifndef TA_TEST_DIR
```
Comment (jforissier) [inline host/xtest/xtest_test.h]: Looks like `TA_TEST_DIR` is not used anymore

Comment (jenswi-linaro) [reply host/xtest/xtest_test.h]: It's still used in `host/xtest/regression_1000.c`. Should I move it back to `host/xtest/regression_1000.c` again?

Comment (jforissier) [reply host/xtest/xtest_test.h]: It's not used in the HEAD of this PR, right?

```
$ git fetch origin pull/240/head
From https://github.com/OP-TEE/optee_test
 * branch                      refs/pull/240/head -> FETCH_HEAD
$ git checkout FETCH_HEAD 
HEAD is now at 80b8e5941579... [review] xtest: add --install-tas argument
$ git grep TA_TEST_DIR
host/xtest/xtest_test.h:#ifndef TA_TEST_DIR
host/xtest/xtest_test.h:#  define TA_TEST_DIR "/data/tee/optee_armtz"
host/xtest/xtest_test.h:#  define TA_TEST_DIR "/tmp/optee_armtz"
```

So I think it should remain in `host/xtest/regression_1000.c` until it finally gets removed by "regression 1008: test with corrupt BSTA".

**Comment by jenswi-linaro** [comment]
Addressed the comments, except the one about `TA_TEST_DIR`.
Rebased and squashed

File: host/xtest/xtest_test.h (hunk 4 above)
Comment (jenswi-linaro) [reply host/xtest/xtest_test.h]: Aha, sorry I confused it with `TA_DIR`.

**Comment by jenswi-linaro** [comment]
Rebased and addressed the `TA_TEST_DIR` stuff too.

**Comment by jbech-linaro** [comment]
How does this affect the AOSP builds? I theory it should just work I guess, but it'd be good to have some confidence that it still works before pushing this. Maybe Yongqin or Vee could be out testing.

File: host/xtest/install_ta.c [hunk 5]
```diff
@@ -0,0 +1,157 @@
+/*
+ * Copyright (c) 2017, Linaro Limited
+ * All rights reserved.
```
Comment (jforissier) [inline host/xtest/install_ta.c]: "All rights reserved" is useless noise. https://reuse.software/practices/.

**Comment by vchong** [comment]
I'll run a test later today and report back.

File: host/xtest/regression_1000.c [hunk 6]
```diff
@@ -31,6 +32,11 @@
 #include <ta_sims_test.h>
 #include <ta_concurrent.h>
 #include <sdp_basic.h>
+#include <pta_secstor_ta_mgmt.h>
```
Comment (jforissier) [inline host/xtest/regression_1000.c]: Missing `#ifdef`

**Comment by vchong** [comment]
Please add changes in `host/xtest/Makefile` to `Android.mk` too. Thanks!

**Comment by vchong** [comment]
Sorry for the delay. Have several build errors to fix. So:

1. `xtest` -> ok

2. `xtest --install-ta /system/lib/optee_armtz` -> ok
```
hikey:/ # xtest --install-ta /system/lib/optee_armtz                                                                   
Searching directory "/system/lib/optee_armtz" for TAs                                                                  
Installing "12345678-5b69-11e4-9dbb-101f74f00099.ta"                                                                   
Installing "484d4143-2d53-4841-3120-4a6f636b6542.ta"                                                                   
Installing "5b9e0e40-2636-11e1-ad9e-0002a5d5c51b.ta"                                                                   
Installing "5ce0c432-0ab0-40e5-a056-782ca0e6aba2.ta"                                                                   
Installing "5dbac793-f574-4871-8ad3-04331ec17f24.ta"                                                                   
Installing "614789f2-39c0-4ebf-b235-92b32ac107ed.ta"                                                                   
Installing "731e279e-aafb-4575-a771-38caa6f0cca6.ta"                                                                   
Installing "873bcd08-c2c3-11e6-a937-d0bf9c45c61c.ta"                                                                   
Installing "8aaaf200-2450-11e4-abe2-0002a5d5c51b.ta"                                                                   
Installing "b689f2a7-8adf-477a-9f99-32e90c0ad0a2.ta"                                                                   
Installing "b6c53aba-9669-4668-a7f2-205629d00f86.ta"                                                                   
Installing "c3f6e2c0-3548-11e1-b86c-0800200c9a66.ta"                                                                   
Installing "cb3e5ba0-adf1-11e0-998b-0002a5d5c51b.ta"                                                                   
Installing "d17f73a0-36ef-11e1-984a-0002a5d5c51b.ta"                                                                   
Installing "e13010e0-2ae1-11e5-896a-0002a5d5c51b.ta"                                                                   
Installing "e626662e-c0e2-485c-b8c8-09fbce6edf3d.ta"                                                                   
Installing "e6a33ed4-562b-463a-bb7e-ff5e15a493c8.ta"                                                                   
Installing "f157cda0-550c-11e5-a6fa-0002a5d5c51b.ta"                                                                   
Installing TAs done
```

3. `xtest` -> ok

4. Repeat 2 and 3 -> ok

5. `xtest --install-ta <foo_uuid>.ta` -> ok

6. `xtest --install-ta <bar_uuid>.ta` -> ok

7. `xtest --install-ta /system/lib/optee_armtz/*.ta`
```
Installing "/system/lib/optee_armtz/12345678-5b69-11e4-9dbb-101f74f00099.ta"                                           
Installing "/system/lib/optee_armtz/484d4143-2d53-4841-3120-4a6f636b6542.ta"                                           
Installing "/system/lib/optee_armtz/5b9e0e40-2636-11e1-ad9e-0002a5d5c51b.ta"                                           
Installing "/system/lib/optee_armtz/5ce0c432-0ab0-40e5-a056-782ca0e6aba2.ta"                                           
Installing "/system/lib/optee_armtz/5dbac793-f574-4871-8ad3-04331ec17f24.ta"                                           
Installing "/system/lib/optee_armtz/614789f2-39c0-4ebf-b235-92b32ac107ed.ta"                                           
Installing "/system/lib/optee_armtz/731e279e-aafb-4575-a771-38caa6f0cca6.ta"                                           
Installing "/system/lib/optee_armtz/873bcd08-c2c3-11e6-a937-d0bf9c45c61c.ta"                                           
Installing "/system/lib/optee_armtz/8aaaf200-2450-11e4-abe2-0002a5d5c51b.ta"                                           
Installing "/system/lib/optee_armtz/b689f2a7-8adf-477a-9f99-32e90c0ad0a2.ta"                                           
Installing "/system/lib/optee_armtz/b6c53aba-9669-4668-a7f2-205629d00f86.ta"                                           
Installing "/system/lib/optee_armtz/c3f6e2c0-3548-11e1-b86c-0800200c9a66.ta"                                           
Installing "/system/lib/optee_armtz/cb3e5ba0-adf1-11e0-998b-0002a5d5c51b.ta"                                           
Installing "/system/lib/optee_armtz/d17f73a0-36ef-11e1-984a-0002a5d5c51b.ta"                                           
Installing "/system/lib/optee_armtz/e13010e0-2ae1-11e5-896a-0002a5d5c51b.ta"                                           
Installing "/system/lib/optee_armtz/e626662e-c0e2-485c-b8c8-09fbce6edf3d.ta"                                           
Installing "/system/lib/optee_armtz/e6a33ed4-562b-463a-bb7e-ff5e15a493c8.ta"                                           
Installing "/system/lib/optee_armtz/f157cda0-550c-11e5-a6fa-0002a5d5c51b.ta"                                           
Installing TAs done
```

8. `xtest` -> ok

**Comment by jenswi-linaro** [comment]
Added (untested) Android patches

**Comment by vchong** [comment]
> Added (untested) Android patches

That's what I have so should be ok. Thanks!

**Comment by jenswi-linaro** [comment]
Do I smell a tested-by tag? ;-)

**Comment by vchong** [comment]
Continued testing from https://github.com/OP-TEE/optee_test/pull/240#issuecomment-349276339

9. `rm -rf /data/tee/*` -> ok

10. `adb push` *.ta files created from an old (but new enough) manifest build to `/system/lib/optee_armtz` and rerun `xtest` -> ok

**Comment by vchong** [comment]
> Do I smell a tested-by tag? ;-)

Oops, forgot. :)

`Tested-by: Victor Chong <victor.chong@linaro.org> (hikey aosp)`

**Comment by jenswi-linaro** [comment]
Squashed and tag applied

File: host/xtest/install_ta.c [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <compiler.h>
+#include <dirent.h>
+#include <err.h>
+#include <errno.h>
+#include <fnmatch.h>
+#include <inttypes.h>
+#include <pta_secstor_ta_mgmt.h>
```
Comment (jforissier) [inline host/xtest/install_ta.c]: Build regression not addressed

Comment (jbech-linaro) [reply host/xtest/install_ta.c]: Don't just read/say that include too quickly :smile:

Comment (jforissier) [reply host/xtest/install_ta.c]: Ha! Got it :wink:

**Comment by jforissier** [comment]
`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

**Comment by jenswi-linaro** [comment]
Tag applied

**Pull request merged**